File: runtime/dart/ui/rect.ts
import { IOffset } from "./offset";
import { ISize } from "./size";
declare const dart: {
    ui: {
        rectFromLTRB: (
            left: number,
            top: number,
            right: number,
            bottom: number
        ) => IRect;
        rectFromLTWH: (
            left: number,
            top: number,
            width: number,
            height: number
        ) => IRect;
        rectFromCircle: (props: { center: IOffset; radius: number }) => IRect;
        rectFromCenter: (props: {
            center: IOffset;
            height: number;
            width: number;
        }) => IRect;
        rectFromPoints: (a: IOffset, b: IOffset) => IRect;
        rectLerp: (
            a: IRect | undefined,
            b: IRect | undefined,
            t: number
        ) => IRect | undefined;
    };
};
export interface IRect {
    left: number;
    top: number;
    right: number;
    bottom: number;
    getWidth: () => number;
    getHeight: () => number;
    getSize: () => ISize;
    getHasNaN: () => boolean;
    getIsInfinite: () => boolean;
    getIsFinite: () => boolean;
    getIsEmpty: () => boolean;
    shift: (offset: IOffset) => IRect;
    translate: (translateX: number, translateY: number) => IRect;
    inflate: (delta: number) => IRect;
    deflate: (delta: number) => IRect;
    intersect: (other: IRect) => IRect;
    expandToInclude: (other: IRect) => IRect;
    overlaps: (other: IRect) => boolean;
    getShortestSide: () => number;
    getLongestSide: () => number;
    getTopLeft: () => IOffset;
    getTopCenter: () => IOffset;
    getTopRight: () => IOffset;
    getCenterLeft: () => IOffset;
    getCenter: () => IOffset;
    getCenterRight: () => IOffset;
    getBottomLeft: () => IOffset;
    getBottomCenter: () => IOffset;
    getBottomRight: () => IOffset;
    contains: (offset: IOffset) => boolean;
    getHashCode: () => number;
    toString: () => string;
}
export class Rect {
    public static zero = Rect.fromLTRB(0.0, 0.0, 0.0, 0.0);
    private static _giantScalar = 1.0e9;
    public static largest = Rect.fromLTRB(
        -Rect._giantScalar,
        -Rect._giantScalar,
        Rect._giantScalar,
        Rect._giantScalar
    );
    public readonly left: number = undefined as any;
    public readonly top: number = undefined as any;
    public readonly right: number = undefined as any;
    public readonly bottom: number = undefined as any;
    public static fromLTRB(
        left: number,
        top: number,
        right: number,
        bottom: number
    ): IRect {
        return dart.ui.rectFromLTRB(left, top, right, bottom);
    }
    public static fromLTWH(
        left: number,
        top: number,
        width: number,
        height: number
    ): IRect {
        return dart.ui.rectFromLTWH(left, top, width, height);
    }
    public static fromCircle(props: {
        center: IOffset;
        radius: number;
    }): IRect {
        return dart.ui.rectFromCircle(props);
    }
    public static fromCenter(props: {
        center: IOffset;
        height: number;
        width: number;
    }): IRect {
        return dart.ui.rectFromCenter(props);
    }
    public static fromPoints(a: IOffset, b: IOffset): IRect {
        return dart.ui.rectFromPoints(a, b);
    }
    public static lerp(
        a: IRect | undefined,
        b: IRect | undefined,
        t: number
    ): IRect | undefined {
        return dart.ui.rectLerp(a, b, t);
    }
    private readonly _dart_getWidth: () => number = undefined as any;
    private readonly _dart_getHeight: () => number = undefined as any;
    private readonly _dart_getSize: () => ISize = undefined as any;
    private readonly _dart_getHasNaN: () => boolean = undefined as any;
    private readonly _dart_getIsInfinite: () => boolean = undefined as any;
    private readonly _dart_getIsFinite: () => boolean = undefined as any;
    private readonly _dart_getIsEmpty: () => boolean = undefined as any;
    private readonly _dart_shift: (offset: IOffset) => IRect = undefined as any;
    private readonly _dart_translate: (
        translateX: number,
        translateY: number
    ) => IRect = undefined as any;
    private readonly _dart_inflate: (delta: number) => IRect = undefined as any;
    private readonly _dart_deflate: (delta: number) => IRect = undefined as any;
    private readonly _dart_intersect: (
        other: IRect
    ) => IRect = undefined as any;
    private readonly _dart_expandToInclude: (
        other: IRect
    ) => IRect = undefined as any;
    private readonly _dart_overlaps: (
        other: IRect
    ) => boolean = undefined as any;
    private readonly _dart_getShortestSide: () => number = undefined as any;
    private readonly _dart_getLongestSide: () => number = undefined as any;
    private readonly _dart_getTopLeft: () => IOffset = undefined as any;
    private readonly _dart_getTopCenter: () => IOffset = undefined as any;
    private readonly _dart_getTopRight: () => IOffset = undefined as any;
    private readonly _dart_getCenterLeft: () => IOffset = undefined as any;
    private readonly _dart_getCenter: () => IOffset = undefined as any;
    private readonly _dart_getCenterRight: () => IOffset = undefined as any;
    private readonly _dart_getBottomLeft: () => IOffset = undefined as any;
    private readonly _dart_getBottomCenter: () => IOffset = undefined as any;
    private readonly _dart_getBottomRight: () => IOffset = undefined as any;
    private readonly _dart_contains: (
        offset: IOffset
    ) => boolean = undefined as any;
    private readonly _dart_getHashCode: () => number = undefined as any;
    private readonly _dart_toString: () => string = undefined as any;
    public getWidth(): number {
        return this._dart_getWidth();
    }
    public getHeight(): number {
        return this._dart_getHeight();
    }
    public getSize(): ISize {
        return this._dart_getSize();
    }
    public getHasNaN(): boolean {
        return this._dart_getHasNaN();
    }
    public getIsInfinite(): boolean {
        return this._dart_getIsInfinite();
    }
    public getIsFinite(): boolean {
        return this._dart_getIsFinite();
    }
    public getIsEmpty(): boolean {
        return this._dart_getIsEmpty();
    }
    public shift(offset: IOffset): IRect {
        return this._dart_shift(offset);
    }
    public translate(translateX: number, translateY: number): IRect {
        return this._dart_translate(translateX, translateY);
    }
    public inflate(delta: number): IRect {
        return this._dart_inflate(delta);
    }
    public deflate(delta: number): IRect {
        return this._dart_deflate(delta);
    }
    public intersect(other: IRect): IRect {
        return this._dart_intersect(other);
    }
    public expandToInclude(other: IRect): IRect {
        return this._dart_expandToInclude(other);
    }
    public overlaps(other: IRect): boolean {
        return this._dart_overlaps(other);
    }
    public getShortestSide(): number {
        return this._dart_getShortestSide();
    }
    public getLongestSide(): number {
        return this._dart_getLongestSide();
    }
    public getTopLeft(): IOffset {
        return this._dart_getTopLeft();
    }
    public getTopCenter(): IOffset {
        return this._dart_getTopCenter();
    }
    public getTopRight(): IOffset {
        return this._dart_getTopRight();
    }
    public getCenterLeft(): IOffset {
        return this._dart_getCenterLeft();
    }
    public getCenter(): IOffset {
        return this._dart_getCenter();
    }
    public getCenterRight(): IOffset {
        return this._dart_getCenterRight();
    }
    public getBottomLeft(): IOffset {
        return this._dart_getBottomLeft();
    }
    public getBottomCenter(): IOffset {
        return this._dart_getBottomCenter();
    }
    public getBottomRight(): IOffset {
        return this._dart_getBottomRight();
    }
    public contains(offset: IOffset): boolean {
        return this._dart_contains(offset);
    }
    public getHashCode(): number {
        return this._dart_getHashCode();
    }
    public toString(): string {
        return this._dart_toString();
    }
}

File: lib/cfr/builtins/libs/dart/ui/rect.dart
import 'dart:core';
import 'dart:ui';

import 'package:hydro_sdk/cfr/runtimeSupport.dart';

class VMManagedRect extends VMManagedBox<Rect> {
  VMManagedRect(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['left'] = vmObject.left;
    table['top'] = vmObject.top;
    table['right'] = vmObject.right;
    table['bottom'] = vmObject.bottom;
    table['getWidth'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.width,
      ];
    });
    table['getHeight'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.height,
      ];
    });
    table['getSize'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Size>(
            object: vmObject.size, hydroState: hydroState, table: HydroTable()),
      ];
    });
    table['getHasNaN'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hasNaN,
      ];
    });
    table['getIsInfinite'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.isInfinite,
      ];
    });
    table['getIsFinite'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.isFinite,
      ];
    });
    table['getIsEmpty'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.isEmpty,
      ];
    });
    table['shift'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Rect>(
            object: vmObject.shift(maybeUnBoxAndBuildArgument<Offset, dynamic>(
                luaCallerArguments[1],
                parentState: hydroState)),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['translate'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Rect>(
            object: vmObject.translate(luaCallerArguments[1]?.toDouble(),
                luaCallerArguments[2]?.toDouble()),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['inflate'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Rect>(
            object: vmObject.inflate(luaCallerArguments[1]?.toDouble()),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['deflate'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Rect>(
            object: vmObject.deflate(luaCallerArguments[1]?.toDouble()),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['intersect'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Rect>(
            object: vmObject.intersect(
                maybeUnBoxAndBuildArgument<Rect, dynamic>(luaCallerArguments[1],
                    parentState: hydroState)),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['expandToInclude'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Rect>(
            object: vmObject.expandToInclude(
                maybeUnBoxAndBuildArgument<Rect, dynamic>(luaCallerArguments[1],
                    parentState: hydroState)),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['overlaps'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.overlaps(maybeUnBoxAndBuildArgument<Rect, dynamic>(
            luaCallerArguments[1],
            parentState: hydroState)),
      ];
    });
    table['getShortestSide'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.shortestSide,
      ];
    });
    table['getLongestSide'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.longestSide,
      ];
    });
    table['getTopLeft'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Offset>(
            object: vmObject.topLeft,
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['getTopCenter'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Offset>(
            object: vmObject.topCenter,
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['getTopRight'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Offset>(
            object: vmObject.topRight,
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['getCenterLeft'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Offset>(
            object: vmObject.centerLeft,
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['getCenter'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Offset>(
            object: vmObject.center,
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['getCenterRight'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Offset>(
            object: vmObject.centerRight,
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['getBottomLeft'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Offset>(
            object: vmObject.bottomLeft,
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['getBottomCenter'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Offset>(
            object: vmObject.bottomCenter,
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['getBottomRight'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Offset>(
            object: vmObject.bottomRight,
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['contains'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.contains(maybeUnBoxAndBuildArgument<Offset, dynamic>(
            luaCallerArguments[1],
            parentState: hydroState)),
      ];
    });
    table['getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hashCode,
      ];
    });
    table['toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.toString(),
      ];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final Rect vmObject;
}

void loadRect({required HydroState hydroState, required HydroTable table}) {
  table['rectFromLTRB'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Rect>(
          object: Rect.fromLTRB(
              luaCallerArguments[1]?.toDouble(),
              luaCallerArguments[2]?.toDouble(),
              luaCallerArguments[3]?.toDouble(),
              luaCallerArguments[4]?.toDouble()),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['rectFromLTWH'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Rect>(
          object: Rect.fromLTWH(
              luaCallerArguments[1]?.toDouble(),
              luaCallerArguments[2]?.toDouble(),
              luaCallerArguments[3]?.toDouble(),
              luaCallerArguments[4]?.toDouble()),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['rectFromCircle'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Rect>(
          object: Rect.fromCircle(
              center: maybeUnBoxAndBuildArgument<Offset, dynamic>(
                  luaCallerArguments.length >= 2
                      ? luaCallerArguments[1]['center']
                      : null,
                  parentState: hydroState),
              radius: luaCallerArguments.length >= 2
                  ? luaCallerArguments[1]['radius']
                  : null?.toDouble()),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['rectFromCenter'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Rect>(
          object: Rect.fromCenter(
              center: maybeUnBoxAndBuildArgument<Offset, dynamic>(
                  luaCallerArguments.length >= 2
                      ? luaCallerArguments[1]['center']
                      : null,
                  parentState: hydroState),
              height: luaCallerArguments.length >= 2
                  ? luaCallerArguments[1]['height']
                  : null?.toDouble(),
              width: luaCallerArguments.length >= 2
                  ? luaCallerArguments[1]['width']
                  : null?.toDouble()),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['rectFromPoints'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Rect>(
          object: Rect.fromPoints(
              maybeUnBoxAndBuildArgument<Offset, dynamic>(luaCallerArguments[1],
                  parentState: hydroState),
              maybeUnBoxAndBuildArgument<Offset, dynamic>(luaCallerArguments[2],
                  parentState: hydroState)),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['rectLerp'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    final returnValue = Rect.lerp(
        maybeUnBoxAndBuildArgument<Rect?, dynamic>(luaCallerArguments[1],
            parentState: hydroState),
        maybeUnBoxAndBuildArgument<Rect?, dynamic>(luaCallerArguments[2],
            parentState: hydroState),
        luaCallerArguments[3]?.toDouble());
    if (returnValue != null) {
      return [
        maybeBoxObject<Rect?>(
            object: returnValue, hydroState: hydroState, table: HydroTable()),
      ];
    }
    return [];
  });
  registerBoxer<Rect>(boxer: (
      {required Rect vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedRect(
        vmObject: vmObject, hydroState: hydroState, table: table);
  });
}
